File: LLD/InterviewExamples/RuleEngine/RuleEngine/lib/services/impl/SimpleRuleEngine.java
package services.impl;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import models.Expense;
import models.ExpenseType;
import services.RuleEngine;
import services.rules.ExpenseRule;
import services.rules.TripRule;
import services.rules.Violation;
public class SimpleRuleEngine implements RuleEngine{
    
    @Override
    public List<Violation> evaluate(   
        List<TripRule> tripRulesRegistry,
        List<Expense> expenses,
        Map<ExpenseType, List<ExpenseRule>> expenseRuleRegistry,
        List<ExpenseRule> allExpenseRuleRegistry
    ) {
        List<Violation> violationsResult = new ArrayList<>();

        // 1. Check all expenses against all expense rules

        for(Expense expense: expenses) {
            List<ExpenseRule> rules = expenseRuleRegistry.getOrDefault(expense.getExpenseType(),new ArrayList<ExpenseRule>());

           
            checkExpenseAgainstRules(expense, rules, violationsResult);
            checkExpenseAgainstRules(expense, allExpenseRuleRegistry, violationsResult);
        }
        
        // 2. Check all expense against all trip rules

        for(TripRule rule: tripRulesRegistry) {
            Optional<Violation> violation = rule.check(expenses);
            if(violation.isPresent()) {
                violationsResult.add(violation.get());
            }
        }

        return violationsResult;
    }

    private void checkExpenseAgainstRules(Expense expense, List<ExpenseRule> rules, List<Violation> violationsResult) {
         for(ExpenseRule rule : rules) {
                Optional<Violation> violation = rule.check(expense);
                if(violation.isPresent()) {
                    violationsResult.add(violation.get());
                }
            }
    }
}

File: LLD/InterviewExamples/RuleEngine/RuleEngineNode/lib/services/impl/SimpleRuleEngine.js
import RuleEngine from "../RuleEngine.js";
export default class SimpleRuleEngine extends RuleEngine {

    evaluate(tripRulesRegistry,expenses,expenseRuleRegistry,allExpenseRuleRegistry) {
        const violationsResult = []

        for(let expense of expenses) {
            let rules = expenseRuleRegistry[expense.expenseType]

            if(Array.isArray(rules)) {
                this.checkExpenseAgainstRules(expense,rules,violationsResult)
            }
            this.checkExpenseAgainstRules(expense, allExpenseRuleRegistry,violationsResult)

        }

        for(let rule of tripRulesRegistry) {
            const violation = rule.check(expenses)
            if(violation) {
                violationsResult.add(violation)
            }
        }

        return violationsResult
    }

    checkExpenseAgainstRules(expense, rules, violationResult) {
        for(let rule of rules) {
            const violation = rule.check(expense)
            if(violation) {
                violationResult.push(violation)
            }
        }
    }
}